manual: automatically generate modules documentation

ericsagnes opened the pull request:
###### Things done
- [ ] Tested using sandboxing
  ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS,
    or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
  - [x] NixOS
  - [ ] OS X
  - [ ] Linux
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

This automatically create manual entries for modules with `meta.doc` declared.

Extra changes:
- addition of `meta.doc` for gitlab and postgresql modules (and indentation tuning)
- fix of taskserver module manual path

cc @edolstra

ericsagnes committed:
manual: automatically generate modules entries

domenkozar commented:
You're my hero - I'll review this shortly.

nbp commented on nixos/modules/services/misc/gitlab.nix:
```diff
@@ -273,179 +273,183 @@ in {
-        openssh
-        nodejs
+  config = mkMerge [
+    (mkIf cfg.enable {
```
@edolstra Time to bring back `mkAlways`?

nbp commented:
The patch looks good to me, except for the addition of the `mkMerge`, but I think this is an issue of the module system where we could give some extra meaning to a top-level `meta` attribute.

ericsagnes commented:
@nbp +1 for a top `meta` attribute, it would make things cleaner and there doesn't seem to be a real need to have `meta` in `config` anyway.

ericsagnes commented:
I made an implementation of the top level attribute approach [here](https://github.com/ericsagnes/nixpkgs/tree/modules/meta-manual).
If everybody think it is better, I can update this PR code with the top level attribute one.

nbp commented:
@ericsagnes 
> […] there doesn't seem to be a real need to have meta in config anyway.

That I agree, but I do not want to add special attributes either.  special attributes tend to diverge and make the integration of modules as sub-modules harder, in general.

So my suggestion was only to have a trick in `unifyModuleSyntax` to forward the `meta` top-level attribute as a `config` option definition.  In which case we should probably also move the option declaration as part of the module system.  (hey who know, maybe one days we would have a full set of documentation for apache extensions too)

Thus, instead of doing a mkMerge in each module, we do it as part of the `unifyModuleSyntax`.

ericsagnes committed:
modules: move meta at top level

ericsagnes commented:
@nbp Sorry for the late reply, I just noticed your comment today.
I added a commit that move the top level `meta` attribute into `config` in `unifyModuleSyntax`.

domenkozar commented:
@ericbmerritt there are conflits for the merge.

@nbp otherwise all good?

ericbmerritt commented:
@ericsagnes I suspect @domenkozar was pinging you not me.

ericsagnes commented:
I noticed it didn't really work like expected while rebasing.
I will put back the WIP in the title and work a little more on this.

ericsagnes commented:
I fixed the conflicts and it should be working as expected now.
@nbp could you review the code, especially the changes to `lib/modules.nix`?

ericsagnes commented:
update: rebased on master to fix conflicts

globin commented:
Any blockers here? @nbp @domenkozar?

ericsagnes commented:
update: 
- rebased on master due to a conflict with nixos-manual module
- fixed a syntax error in wpa_supplicant module that was breaking manual generation

For the record, a side effect of this PR is to facilitate the ownership declaration of modules by moving `meta.maintainers` at top level. (no need for `mkMerge` anymore)

nbp commented:
@ericsagnes Nothing more to say besides **Thanks You** !

domenkozar commented:
I'm +1 for merging this. Let's wait for @edolstra and merge.

ericsagnes commented:
@nbp Thanks for reviewing!
I fixed the recent conflicts with master, so this should be ready to merge.

domenkozar commented:
No feedback from @edolstra, since this is very minor in terms of code but big improvements for NixOS modules declarative configuration, I'm going to merge it.
